File: domain-persistence-modeling/java-spring-boot/src/main/java/x/scratch/TestListener.java
package x.scratch;

import org.springframework.context.ApplicationListener;
import org.springframework.stereotype.Component;

import javax.annotation.Nonnull;
import java.util.ArrayList;
import java.util.List;

@Component
public class TestListener<Event extends DomainChangedEvent<?>>
        implements ApplicationListener<Event> {
    private final List<Event> received = new ArrayList<>();

    @Override
    public void onApplicationEvent(@Nonnull final Event event) {
        received.add(event);
    }

    public void reset() {
        received.clear();
    }

    public List<Event> events() {
        final var events = new ArrayList<>(received);
        reset();
        return events;
    }
}

File: domain-persistence-modeling/kotlin-spring-boot/src/test/kotlin/x/domainpersistencemodeling/TestListener.kt
package x.domainpersistencemodeling

import ch.tutteli.atrium.creating.ReportingAssertionPlant
import ch.tutteli.atrium.verbs.expect
import org.springframework.context.ApplicationListener
import org.springframework.stereotype.Component

/** @todo Do not listen for all domain changed events, just [Event] ones */
@Component
class TestListener<Event : DomainChangedEvent<*>>
    : ApplicationListener<Event> {
    private val received = mutableListOf<Event>()

    /** Creates a `List` expectation for Atrium, and resets the listener. */
    val expectNext: ReportingAssertionPlant<List<Event>>
        get() = expect(received.toList()).also {
            reset()
        }

    override fun onApplicationEvent(event: Event) {
        received.add(event)
    }

    fun reset() = received.clear()

    @Suppress("unused")
    fun dump() = println(received)
}
